File: 123-best-time-to-buy-and-sell-stock-iii.py


class Solution:

    def maxProfit(self, prices):
        """
        :type k: int
        :type prices: List[int]
        :rtype: int
        """
        
        k = 2
        n = len(prices)
        if n <= 1: return 0

        dp, m = [], []
        for t in range(0, k+1):
            dp.append(0)
            m.append(-prices[0])

        for i in range(1, n):
            for t in range(k, 0, -1):
                m[t] = max(m[t], dp[t-1] - prices[i-1])
                dp[t] = max(dp[t], prices[i] + m[t])
        
        return dp[k]


    # --------------------------------------------- 

    def maxProfit_TimeLimitExceeded(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        
        if len(prices) <= 1:
            return 0
        
        maximum = 0
        for i in range(len(prices)):
            p1 = self.maxOneTransaction(prices[:i])
            p2 = self.maxOneTransaction(prices[i:])
            p = p1 + p2
            if maximum < p:
                maximum = p
        return maximum
        

    def maxOneTransaction(self, prices):
        
        if len(prices) <= 1:
            return 0
        
        profit = 0
        minPrice = prices[0]
        for price in prices:
            prof = price - minPrice
            if prof > profit:
                profit = prof
            if price < minPrice:
                minPrice = price
        return profit

File: 123-best-time-to-buy-and-sell-stock-iii.java


class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        if (n <= 1) return 0;

        int[] left = new int[n];
        int[] right = new int[n];

        // left side
        int min = prices[0];
        left[0] = 0;
        for (int i = 1; i < n; i++) {
            int p = 0;
            if (prices[i] > min) p = prices[i] - min;
            left[i] = Math.max(left[i-1], p);

            min = Math.min(min, prices[i]);
        }

        // right side
        int max = prices[n-1];
        right[n-1] = 0;
        for (int i = n-2; i >= 0; i--) {
            int p = 0;
            if (prices[i] < max) p = max - prices[i];
            right[i] = Math.max(right[i+1], p);

            max = Math.max(max, prices[i]);
        }

        int profit = left[n-1];
        for (int i = 0; i < n-1; i++) {
            profit = Math.max(profit, left[i]+right[i+1]);
        }

        return profit;
    }
}